src/Grammar.java

```diff
@@ -0,0 +1,91 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
+
+public class Grammar {
+    private List<String> nonTerminals;
+    private Set<String> terminals;
+    private List<Production> productions;
+    private String filename;
+
+    public Grammar(String filename) {
+        nonTerminals = new LinkedList<>();
+        terminals = new HashSet<>();
+        productions = new ArrayList<>();
+        this.filename = filename;
+        getGrammarFromFile();
+    }
+
+    private void getGrammarFromFile() {
+        try {
+            int i = 0;
+            for (String line : Files.readAllLines(Paths.get(this.filename))) {
+                if (i < 2){
+                    String[] tokens = line.split(" ");
+                    for (String token : tokens) {
+                        if (i == 0)
+                            nonTerminals.add(token);
+                        if (i == 1)
+                            terminals.add(token);
+                    }
+                }
+
+                if (i >= 2) {
+                    String[] tokens = line.split(" - ");
+                    List<List<String>> rules = new ArrayList<>();
+
+                    for ( String rule: tokens[1].split(" \\| "))
+                        rules.add(Arrays.asList(rule.split(" ")));
+                    productions.add(new Production(tokens[0], rules));
+                }
+                i++;
+            }
+        } catch (IOException e) {
```

Comment:
It's fine like this, but maybe throw the exception again, not only log it?

This way it will stop the execution and the issue will not be lost in the logs, in which case it can lead to other issues and you will not see the primary one, just the one generated by it.